queue.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+#include <stdexcept>
+using namespace std;
+
+template <typename T1>
+class BasicQueue
+{
+public:
+
+    // Add element to the end
+    virtual void enqueue(T1 &) = 0;
+
+    // Return first element
+    virtual T1 dequeue() = 0;
+
+    // Get total count of elements
+    virtual size_t getLength() = 0;
+
+};
+
+
+
+template <typename T1>
+class ListQueue: public BasicQueue<T1>
+{
+private:
+	template <typename T1>
+	class node
+	{
+	public:
+		node(T1 x) : data(x), prev(nullptr) {};
+
+		T1 data;
+		node * prev;
+	};
+
+	node<T1> * head, *tail;
+	size_t size;
+
+public:
+	ListQueue() : head(nullptr), tail(nullptr), size(0) {};
+	ListQueue(node<T1> * x) : head(x), tail(x), size(1) {};
+	ListQueue(ListQueue<T1> & l) : head(l.head), tail(l.tail), size(l.size) {};

```

Comment:
Here should be destructor. Currently, deleting non-empty ListQueue leads to memory leaks.
Also, I want to notice that your copy constructor is very dangerous. Furthermore, I think you should delete it. Problem is that this is not "deep" copy, resulting in both objects pointing on the same data.